src/services/users.ts

```diff
@@ -54,9 +65,86 @@ export function getUserAttributes(dbPool: PostgresJsDatabase<typeof db>) {
   };
 }
 
+/**
+ * Checks user data for existing entries in the database.
+ * @param {PostgresJsDatabase<typeof db>} dbPool - The database connection pool.
+ * @param {string} userId - The ID of the user to check.
+ * @param {UserData} userData - The user data to check.
+ * @returns {Promise<Array<string> | null>} - An array of errors if user data conflicts, otherwise null.
+ */
+async function checkUserData(
+  dbPool: PostgresJsDatabase<typeof db>,
+  userId: string,
+  userData: UserData,
+) {
+  if (userData.email || userData.username) {
+    const existingUser = await dbPool
+      .select()
+      .from(db.users)
+      .where(
+        or(
+          and(eq(db.users.email, userData.email ?? ''), ne(db.users.id, userId)),
+          and(eq(db.users.username, userData.username ?? ''), ne(db.users.id, userId)),
+        ),
+      );
+
+    if (existingUser.length > 0) {
+      const errors = [];
+
+      if (existingUser[0]?.email && existingUser[0].email === userData.email) {
+        errors.push('Email already exists');
+      }
+
+      if (existingUser[0]?.username && existingUser[0].username === userData.username) {
+        errors.push('Username already exists');
+      }
+
+      return errors;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Upserts user data in the database.
+ * @param {PostgresJsDatabase<typeof db>} dbPool - The database connection pool.
+ * @param {string} userId - The ID of the user to update.
+ * @param {UserData} userData - The updated user data.
+ * @returns {Promise<any>} - A promise resolving to the updated user data.
```

Comment:
lets remove this <any> in the tsdoc

Reply:
either we leave let typescript infer the return value or we write what it actually returns but the any is worse for a user of this function